torch/csrc/distributed/autograd/utils.cpp

```diff
@@ -116,23 +116,23 @@ std::shared_ptr<FutureMessage> sendMessageWithAutograd(
     const WorkerInfo& dst,
     torch::distributed::rpc::Message&& wrappedRpcMsg,
     bool forceGradRecording,
-    const std::shared_ptr<torch::autograd::profiler::RecordFunction>& rf) {
+    const std::shared_ptr<torch::autograd::profiler::RecordFunctionAsync>& rf) {
   auto msg = getMessageWithAutograd(
       dst.id_,
       std::move(wrappedRpcMsg),
       MessageType::FORWARD_AUTOGRAD_REQ,
       forceGradRecording);
 
+  if (rf != nullptr) {
```

Comment:
Instead of calling `exitScope` explicitly, can we create a RecordFunctionAsyncGuard that wraps RecordFunctionAsync? Then in rpc/init.cpp wherever we have RecordFunctionAsync, we create this guard by wrapping RecordFunctionAsync and the destructor of the guard calls `exitScope`.

Taking this a step further, we can make `exitScope` private and only allow RecordFunctionAsyncGuard to call it (as a friend class).